src/RedisSubscriber.js

```diff
@@ -48,11 +52,12 @@ export default class RedisSubscriber implements Subscriber {
     if (!channelQueues) {
       channelQueues = new Set();
       this._queues.set(channel, channelQueues);
-      this.redis.subscribe(channel);
+      await this._redisSubscribe(channel);
     }
 
+    await this._subscribeToChannel(channel);
+
     const queue = new AsyncQueue({
-      setup: () => this._subscribeToChannel(channel),
       teardown: () => {
```

Comment:
is it safe to not try-catch wrap the subscribe, and leave teardown to here?

Reply:
I think so, because the teardown matters only if there is an actual queue

Reply:
what's a good strategy to test this?

Reply:
no, the teardown needs to undo the setup... this isn't obviously correct to me. better to move the try-catch outside i think

Reply:
no, wait, that's not quite right either...

i think the right approach is to modify `AsyncQueue` so that `AsyncQueue` itself returns `Promise<AsyncGenerator<*>>`, and does the awaiting of `setup`

Reply:
Yeah this isn't correct now that i'm looking closer. The setup needs to happen in the context of the generator otherwise it may not be teared down since the generator may not have started (and `return()` isn't called).

AsyncQueue already awaits `setup` isn't it enough to return a promise from `subscribeToChannel` for what this is trying to fix?

Reply:
Instead of an async iterable, we want a promise for an async iterable that resolves when `setup` finishes running.

The idea is that, if we want to publish an initial message, the right time to publish the initial message is immediately after Redis acknowledges our `subscribe` call. Otherwise there's a slight race condition in the gap between when we send the `subscribe` call and when it comes back.

Reply:
the problem tho is you really can't move the setup call here it needs to be lazy with the first `await` of the iterator, otherwise it won't clean up if the subscription is killed before a message is sent. You can't return a `Promise<AsyncIterator>` where the promise resolves after redis subscribes, because that can't run until you begin iterating. 

Best thing I can think of here is a `ack` function for when `setup` finishes that you can pass into `subscribe`...

Reply:
```js
subscribe('foo', parseMessage, () => {
  fireTheMissiles()
})
```

Reply:
or we multiple return values

Reply:
```js
async function* foo() {
  yield 1;
  yield 2;
  yield 3;
}

async function bar() {
  const stream = foo();
  const fv = await stream.next();
  console.log('fv', fv);
  return stream;
}

async function baz() {
  for await (const f of await bar()) {
    console.log('value', f);
  }
}

baz();
```

i think this works

this API sucks, is this seriously shipping as part of JS?

Reply:
Add some `await yield*`s for good measure :P

I don't actually think the ack is that bad

Reply:
yeah but a promise for an async iterator is better... and i think what i have works

Reply:
The above tho will resolve on the first event tho not the setup unless we are gonna yield the setup which is weird generally no?

Reply:
so it'd look something like:

```js
async function createIterable() {
  const iterableRaw = createIterableRaw();

  const initialValue = await iterableRaw.next();
  invariant(initialValue === INITIAL_VALUE, 'unexpected initial value');

  return iterableRaw;
}

async function* createIterableRaw() {
  try {
    await setup();
    yield INITIAL_VALUE;

    yield* createActualIterable();
  } finally {
    teardown();
  }
}
```

Reply:
Do you think that is a better api tho than an optional `ack` callback? I'm not sure how we'd have this be mostly transparent to the callers of `subscribe` that don't care about that first value, and still have it safe to pass through to `filter` or `map` operations. Remember also at the Queue is not 1-1 with `subscribe` calls, setup may have happened already for some consumers how would those ones know that setup is already done, you'd have to tag the queue like `isSetup` and check that before trying to wait on a value that may be an event

Reply:
The idea is that the output wouldn't be safe to immediately pass through to `filter` or `map`. The user doesn't get an async iterable with an extra event – he or she gets a promise for an async iterable that he or she must then `await`. In that sense, then, it's transparent, so the code looks like:

```js
const iterable = await queue.iterable;
await doStuffAfterSetup();
yield* iterable;
```

Reply:
Ah right ok it's hidden in the subscriber. That's not so bad I guess then.

Reply:
yeah the tricky thing is if the client tries to unsubscribe before setup is done... but i think we already deal with that anyway?